verification/spec/connection-handshake/ConnectionHandshakeModule.tla

```diff
@@ -0,0 +1,231 @@
+--------------------- MODULE ConnectionHandshakeModule ---------------------
+
+EXTENDS Naturals, FiniteSets, Sequences, TLC
+
+
+CONSTANTS MaxHeight,        \* Maximum height of the local chain.
+          ConnectionIDs,    \* The set of all possible connection IDs.
+          ClientIDs,        \* The set of all possible client IDs.
+          MaxBufLen         \* Maximum length of the input and output buffers.
+
+
+ASSUME Cardinality(ConnectionIDs) >= 1
+ASSUME Cardinality(ClientIDs) >= 1
+
+
+VARIABLES
+    inBuf,      \* A buffer (Sequence) holding any message(s) incoming to the chModule.
+    outBuf,     \* A buffer (Sequence) holding outbound message(s) from the chModule.
+    store       \* The local store of the chain running this chModule. 
+
+
+ConnectionEnds ==
+    [
+        connectionID : ConnectionIDs,
+        clientID : ClientIDs
+        \* commitmentPrefix add later
+    ]
+
+
+nullConnection == [state |-> "UNINIT"]
+
+
+Heights == 1..MaxHeight
+
+
+Clients ==
+    [   
+        clientID : ClientIDs,
+        consensusState : Heights
+    ]
+
+nullClient == "no client"
+
+
+(* The set of all types which a message can take. 
+    
+    TODO: These types are also used in the Environment module.
+    Is it possible to restrict these to a single module? 
+ *)
+CHMessageTypes ==
+    { "CHMsgInit", "CHMsgTry", "CHMsgAck", "CHMsgConfirm"}
+
+
+(***************************************************************************
+ Helper operators.
+ ***************************************************************************)
+
+(* Returns true if 'para' matches the parameters in the local connection,
+    and returns false otherwise.
+ *)
+CheckLocalParameters(para) ==
+     /\ store.connection.parameters.localEnd.connectionID = para.localEnd.connectionID   
+     /\ store.connection.parameters.remoteEnd.connectionID = para.remoteEnd.connectionID
+     /\ store.connection.parameters.localEnd.clientID = para.localEnd.clientID       
+     /\ store.connection.parameters.remoteEnd.clientID = para.remoteEnd.clientID
+
+
+(* Validates a ConnectionParameter.
+    Expects as input a ConnectionParameter 'para' and returns true or false.
+    
+    This is a basic validation step, making sure that 'para' is valid with
+    respect to module-level constants ConnectionIDs and ClientIDs.
+    If there is a connection in the store, this also validates 'para'
+    against the parameters of an existing connection by relying on the
+    state predicate CheckLocalParameters.
+*)
+ValidConnectionParameters(para) ==
+    /\ para.localEnd.connectionID \in ConnectionIDs
```

Comment:
These kind of checks are normal part of TypeOK if I am not wrong.

Reply:
This is a module-specific check (not a global type invariant, which is captured in TypeOK). In this validation check, we make sure that the connection parameters are valid with respect to the connections and clients that this module expects.

Reply:
In that case, \in conditions need to be more precise as valid connection and client ids are chain specific.